src/CloudNative.CloudEvents/CloudEvent.cs

```diff
@@ -1,123 +1,202 @@
+                    {
+                        throw new ArgumentException("Cannot add an unknown non-extension attribute to an event.");
+                    }
+                    extensionAttributes[name] = attribute;
+                }
+
+                if (value is null)
                 {
-                    extensionMap.Add(extension.GetType(), extension);
-                    extension.Attach(this);
+                    attributeValues.Remove(name);
+                    return;
                 }
+                // TODO: We could convert the attribute value here instead? Or is that a bit too much "magic"?
+                attributeValues[name] = attribute.Validate(value);
             }
-            Extensions = new ReadOnlyDictionary<Type, ICloudEventExtension>(extensionMap);
         }
 
         /// <summary>
-        /// CloudEvent 'datacontenttype' attribute. Content type of the 'data' attribute value.
-        /// This attribute enables the data attribute to carry any type of content, whereby
-        /// format and encoding might differ from that of the chosen event format.
+        /// Sets or fetches the value associated with the given attribute name.
+        /// Setting a value of null removes the value from the event, if it exists.
+        /// If the attribute is not known in this event, fetching the value always returns null, and
+        /// setting the value add a new extension attribute with the given name, and a type of string.
+        /// (The value for an unknown attribute must be a string or null.)
         /// </summary>
-        /// <see cref="https://github.com/cloudevents/spec/blob/master/spec.md#contenttype"/>
-        public ContentType DataContentType
+        public object this[string attributeName]
         {
-            get => attributes[CloudEventAttributes.DataContentTypeAttributeName(attributes.SpecVersion)] as ContentType;
-            set => attributes[CloudEventAttributes.DataContentTypeAttributeName(attributes.SpecVersion)] = value;
+            // TODO: Validate the attribute name? Check it's not spec version?
+            get => attributeValues.GetValueOrDefault(Preconditions.CheckNotNull(attributeName, nameof(attributeName)));
+            set
+            {
+                Preconditions.CheckNotNull(attributeName, nameof(attributeName));
+                var knownAttribute = GetAttribute(attributeName);
+
+                // TODO: Decide on spec version behavior
+
+                // TODO: Are we happy to add the extension in even if the value is null?
+                if (knownAttribute is null)
+                {
+                    if (value is object && !(value is string))
+                    {
+                        throw new ArgumentException($"Cannot assign value of type {value.GetType()} to unknown attribute '{attributeName}'");
+                    }
+                    knownAttribute = CloudEventAttribute.CreateExtension(attributeName, CloudEventAttributeType.String);
```

Comment:
Why is this always a string? If the user hasn't registered the attribute, we just assume string for simplicity and safety?

Reply:
Yes. Every CloudEvent attribute type has a canonical string representation, so this feels like a reasonable balance between safety and convenience. (An alternative would be to reject it entirely - which would be a reasonable choice, forcing callers to create the extension attribute themselves and use the other indexer if they wanted to.) Worth revisiting after writing conceptual docs.